Multidimensional sample

mmann1123 opened the pull request:
Adding ability to handle multidimensional sample by prestacking along those dims.  Also trying to clarify documentation (for the arg I actually think I understand)

NOTES: I think this is working well for DataArrays but not sure about DataSets since I don't know what they should look like. 
PASSES TESTS etc

referencing issue #52

Make sure you have:
- [ ] added the necessary updates to `doc/content/whatsnew.rst`
- [ ] added documentation for new features
- [ ] run bumpversion and reset conda build number if you are preparing a new release

mmann1123 committed:
Multidimensional sample

Adding ability to handle multidimensional sample by prestacking along those dims. PASSES TESTS etc
NOTE: I think this is working well for DataArrays but not sure about DataSets since I don't know what they should look like. 

referencing issue #52

phausamann commented:
Thanks @mmann1123, although I would probably go another route with this. The docstring clarifications are definitely necessary, and I'd probably add a paragraph to the user guide explaining what sample and feature dimensions refer to. 

In any case, the `Featurizer` is used when you have a single sample dimension and everything else is considered a feature dimension, e.g. if you have raster data with dimensions `(time, x, y, band)` and you want to predict something across the whole image for each timestep then `time` would be the sample dim and the rest would be your feature dims.

In your case, it's the other way around, there's only one feature dim (`band`) and the rest are sample dims. I would probably add a dedicated transformer (`Samplerizer`?) for that, or maybe create more general `Stacker`/`Unstacker` transformers.

If you want to give that a try go ahead, otherwise I can also take care of it.

mmann1123 commented:
Hey @phausamann I can take a try. Although I have to say I am still relatively new to python so I can't make any promises.

mmann1123 committed:
attempt at stackerizer

This appears to be working on dataarray not sure how to test on datasets. Also not sure how to handle groupby in BaseTransformer

phausamann commented:
Looks pretty good already. You should add a test in `tests/test_preprocessing.py` though, especially since you've come across a case where it doesn't work in a pipeline. You can probably use `test_featurize` as a guide for writing such a test. I'll also leave a couple of more specific remarks as review comments.

phausamann commented on sklearn_xarray/preprocessing.py:
```diff
@@ -1593,3 +1613,134 @@ def reduce(X, return_estimator=False, **fit_params):
+
+    
+ 
+class Stackerizer(BaseTransformer):
```
I think `Stacker` would be a more fitting name
```suggestion
class Stacker(BaseTransformer):
```

phausamann commented on sklearn_xarray/preprocessing.py:
```diff
@@ -1593,3 +1613,134 @@ def reduce(X, return_estimator=False, **fit_params):
+        self,
+        stack_dims=None,
+        direction="stack",
+        sample_dim ="sample",
```
I think it might make sense to make this a little more flexible, the multi-index dim doesn't necessarily have to be the sample dim:
```suggestion
        new_dim="sample",
```

phausamann commented on sklearn_xarray/preprocessing.py:
```diff
@@ -1593,3 +1613,134 @@ def reduce(X, return_estimator=False, **fit_params):
+        stack_dims=None,
+        direction="stack",
+        sample_dim ="sample",
+        transposed = True,
```
```suggestion
        transposed = False,
```

phausamann commented on sklearn_xarray/preprocessing.py:
```diff
@@ -1593,3 +1613,134 @@ def reduce(X, return_estimator=False, **fit_params):
+
+    
+
+def stackerizer(X, return_estimator=False, **fit_params):
```
```suggestion
def stack(X, return_estimator=False, **fit_params):
```

mmann1123 replied:
any conflict since there is a stack method for xarray? (again new to this level programming) Honestly I'll leave that to you, but just a thought.

phausamann replied:
That's the nice thing about Python: even if these two methods have the same name, they are in different namespaces (`sklearn_xarray.preprocessing.stack` vs `xarray.DataArray.stack`), so the risk of unintended conflicts is very low.

phausamann commented on sklearn_xarray/preprocessing.py:
```diff
@@ -1298,8 +1305,10 @@ def _transform_var(self, X):
+            if isinstance(self.sample_dim, str):
+                self.sample_dim = [self.sample_dim]
+            stack_dims = tuple(set(X.dims) - set(self.sample_dim))
+            
```
These changes should probably be reverted now that there's a dedicated transformer for this

mmann1123 replied:
Ok not sure how to handle this. Can i leave it to you?

phausamann replied:
Don't worry, if you feel that this is too much at once I can take it from here.

mmann1123 replied:
Sorry yeah not quite sure what you're looking for.

mmann1123 committed:
Update sklearn_xarray/preprocessing.py

Co-authored-by: Peter Hausamann <phausamann@users.noreply.github.com>

mmann1123 committed:
Update sklearn_xarray/preprocessing.py

Co-authored-by: Peter Hausamann <phausamann@users.noreply.github.com>

mmann1123 committed:
Update sklearn_xarray/preprocessing.py

Co-authored-by: Peter Hausamann <phausamann@users.noreply.github.com>

phausamann committed:
Update sklearn_xarray/preprocessing.py

phausamann commented:
I've added a slightly modified Stacker transformer to the develop branch (#54), so I'll be closing this PR if that's okay. I've also added a new section to the docs that explains sample and feature dimensions and how to pre-process xarray data such that it has the dimensionality that sklearn expects.
